app/Http/Controllers/Webapp/PostController.php

```diff
@@ -0,0 +1,1890 @@
+            $submitType = 'draft';
+        }
+
+        if($submitType == 'publish' && $this->currentUser->is(User::CONTRIBUTOR)){
+            $this->activityLog->setInfo('Contributor can not publish');
+            return '404 Not Found';
+        }
+
+        $post->save();
+
+        $post->saveHistory();
+
+        $this->activityLog->setPostId($post->id);
+        $this->activityLog->addMeta('SubmitType', $submitType);
+
+        if($submitType == 'publish'){
+            try {
+                if (!Post::isUniquePostByTitle($post)) {
+                    throw new TitleDuplicatedException('このタイトルは存在しています。公開できないのでタイトルを変更してください。');
+                }
+
+                if (!$this->currentUser->is(User::CONTRIBUTOR) && $post->site_id > 0) {
+                    $apiUrl = Site::find($post->site_id)->api_url;
+                    $returnData = (new CdsApiDriver($apiUrl, [
+                        'content' => $post->content,
+                        'title' => $postTitle,
+                        'categories' => [$post->category],
+                        'post_excerpt' => $postDescription,
+                        'post_status' => $submitType,
+                        'feature_img' => $post->feature_img,
+                        'author_name' => !empty($author = User::find($post->user_id)) ? $author->username : '',
+                        'cback_url' => action('Webapp\\PostController@getEdit', $post->id),
+                        'priority' => $post->priority,
+                    ]))->send('publish')->getResult();
+
+                    if ($returnData->data->post_id == 0) {
+                        $this->activityLog->addMeta('PublishedStatus', 'failed');
+                        throw new \Exception();
+                    }
+                    $post->content = Post::replaceIDButtonLink($returnData->data->post_id, $post->content);
+                    $post->target_post_id = $returnData->data->post_id;
+                    $post->published_status = 1;
+                    $this->activityLog->addMeta('PublishedStatus', 'success');
+                    $this->activityLog->addMeta('TargetPostId', $post->target_post_id);
+                }
+
+                $this->_flashMessage($request, '公開しました', 'success');
+            } catch(TitleDuplicatedException $e) {
+                $this->_flashMessage($request, $e->getMessage(), 'danger');
+                $this->activityLog->addMeta('Error', 'Title duplicated: '.$post->title);
+                $post->title = $oldTitle;
+            } catch(\Exception $e) {
+                $post->published_status = 0;
+                $this->activityLog->addMeta('PublishedStatus', 'failed');
+                $this->_flashMessage($request, '更新に失敗しました。もう一度ご確認ください。', 'warning');
+            }
+
+            $post->save();
+        }else{
+            $this->_flashMessage($request, '下書きに保存しました', 'success');
+        }
+
+        $this->activityLog->addMeta('NewPublishedStatus', $post->published_status);
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "NewPublishedStatus" 4 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS1192)